File: src/main/java/ar/com/flow/bankaccount/transaction/withdrawal/Withdrawal.java
package ar.com.flow.bankaccount.transaction.withdrawal;

import ar.com.flow.bankaccount.BankAccount;
import ar.com.flow.bankaccount.transaction.Action;
import ar.com.flow.bankaccount.transaction.Step;
import ar.com.flow.bankaccount.transaction.receipt.DebitStamper;
import ar.com.flow.bankaccount.transaction.Transaction;
import ar.com.flow.bankaccount.transaction.receipt.ReceiptStamper;
import ar.com.flow.money.Money;

public class Withdrawal {
    public static WithdrawalBuilder from(BankAccount debitAccount) {
        return new WithdrawalBuilder(debitAccount);
    }

    public static class WithdrawalBuilder {
        private final BankAccount debitAccount;

        public WithdrawalBuilder(BankAccount debitAccount) {
            this.debitAccount = debitAccount;
        }

        public Transaction amount(Money amountToWithdraw) {
            return Transaction.builder()
                    .amount(amountToWithdraw)
                    .precondition(new SufficientFunds(debitAccount, amountToWithdraw))
                    .step(new Step(debitAccount, Withdrawal.receipt(debitAccount)))
                    .build();
        }
    }

    public static ReceiptStamper receipt(BankAccount account) {
        return new DebitStamper(account, Action.Withdrawal);
    }
}

File: src/main/kotlin/ar/com/flow/bankaccount/domain/Withdrawal.kt
package ar.com.flow.bankaccount.domain

import ar.com.flow.bankaccount.domain.transaction.Builder
import ar.com.flow.bankaccount.domain.transaction.Transaction
import ar.com.flow.bankaccount.domain.transaction.receipt.*
import ar.com.flow.bankaccount.domain.transaction.steps.Step
import ar.com.flow.bankaccount.domain.withdrawal.SufficientFunds
import ar.com.flow.money.Money

internal object Withdrawal {
    fun from(debitAccount: BankAccount): WithdrawalBuilder {
        return WithdrawalBuilder(debitAccount)
    }

    internal class WithdrawalBuilder(private val debitAccount: BankAccount) {
        fun amount(amountToWithdraw: Money): Transaction {
            return Builder()
                .amount(amountToWithdraw)
                .precondition(
                    SufficientFunds(
                        debitAccount,
                        amountToWithdraw
                    )
                )
                .step(Step(debitAccount, WithdrawalDebitPrint(debitAccount), WithdrawalDebitScratch(debitAccount)))
                .build()
        }
    }
}